File: frontend/components/api_client.py
import requests
import streamlit as st
from typing import Dict, List, Optional, Any
from config import settings

class CMBClusterAPIClient:
    """API client for CMBCluster backend"""
    
    def __init__(self, base_url: str = None):
        self.base_url = (base_url or settings.api_url).rstrip('/')
        self.session = requests.Session()
        # Add timeout to prevent hanging requests
        self.session.timeout = 10
    
    def _get_headers(self) -> Dict[str, str]:
        """Get headers with authentication token"""
        headers = {"Content-Type": "application/json"}
        token = st.session_state.get("access_token")
        if token:
            headers["Authorization"] = f"Bearer {token}"
        return headers
    
    def _handle_response(self, response: requests.Response) -> Dict[str, Any]:
        """Handle API response and errors"""
        try:
            response.raise_for_status()
            return response.json()
        except requests.exceptions.HTTPError as e:
            if response.status_code == 401:
                st.error("Authentication failed. Please login again.")
                st.session_state.authenticated = False
                raise e
            elif response.status_code == 403:
                st.error("Access denied. Insufficient permissions.")
                raise e
            elif response.status_code == 404:
                # For 404, return error response instead of empty response
                try:
                    error_data = response.json()
                    return {"status": "error", "message": error_data.get("detail", "Not found")}
                except:
                    return {"status": "error", "message": "Resource not found"}
            elif response.status_code == 500:
                try:
                    error_data = response.json()
                    return {"status": "error", "message": error_data.get("detail", "Server error")}
                except:
                    return {"status": "error", "message": "Server error. Please try again later."}
            else:
                try:
                    error_data = response.json()
                    return {"status": "error", "message": error_data.get("detail", f"API error: {response.status_code}")}
                except:
                    return {"status": "error", "message": f"API error: {response.status_code}"}
        except requests.exceptions.Timeout:
            return {"status": "error", "message": "Request timeout. Please check your connection."}
        except requests.exceptions.ConnectionError:
            return {"status": "error", "message": "Connection error. Please check if the backend is running."}
        except requests.exceptions.RequestException as e:
            return {"status": "error", "message": f"Network error: {str(e)}"}
    
    def create_environment(self, config: Optional[Dict] = None) -> Dict[str, Any]:
        """Create a new user environment"""
        url = f"{self.base_url}/environments"
        
        # Default config
        data = {
            "cpu_limit": 1.0,
            "memory_limit": "2Gi",
            "storage_size": "10Gi"  # Kept for backward compatibility
        }
        
        if config:
            data.update(config)
            
            # Handle storage selection conversion
            if 'storage' in config:
                storage_info = config['storage']
                
                if storage_info.get('selection_type') == 'existing':
                    # Use existing storage
                    data['storage_id'] = storage_info.get('storage_id')
                    data['create_new_storage'] = False
                elif storage_info.get('selection_type') == 'create_new':
                    # Create new storage
                    data['storage_id'] = None
                    data['create_new_storage'] = True
                    if 'storage_class' in storage_info:
                        data['storage_class'] = storage_info['storage_class']
                else:
                    # Default behavior - let backend handle it
                    data['storage_id'] = None
                    data['create_new_storage'] = False
                
                # Remove the frontend 'storage' key
                del data['storage']
        
        response = self.session.post(url, json=data, headers=self._get_headers(), timeout=30)
        return self._handle_response(response)
    
    def get_environment_status(self) -> Dict[str, Any]:
        """Get current environment status"""
        url = f"{self.base_url}/environments"
        response = self.session.get(url, headers=self._get_headers(), timeout=10)
        return self._handle_response(response)
    

    def delete_environment(self, env_id: str = None) -> Dict[str, Any]:
        """Delete user environment. If env_id is provided, delete that environment."""
        url = f"{self.base_url}/environments"
        params = {"env_id": env_id} if env_id else None
        
        response = self.session.delete(url, headers=self._get_headers(), params=params, timeout=20)
        return self._handle_response(response)
    
    def send_heartbeat(self) -> Dict[str, Any]:
        """Send heartbeat to keep environment alive"""
        url = f"{self.base_url}/environments/heartbeat"
        response = self.session.post(url, headers=self._get_headers(), timeout=5)
        return self._handle_response(response)
    
    def get_activity_log(self, limit: int = 50) -> Dict[str, Any]:
        """Get user activity log"""
        url = f"{self.base_url}/activity"
        params = {"limit": limit}
        response = self.session.get(url, params=params, headers=self._get_headers())
        return self._handle_response(response)
    
    def health_check(self) -> Dict[str, Any]:
        """Check API health"""
        url = f"{self.base_url}/health"
        response = self.session.get(url)
        return self._handle_response(response)

    def list_environments(self) -> Dict[str, Any]:
        """List all environments for the current user"""
        url = f"{self.base_url}/environments/list"
        response = self.session.get(url, headers=self._get_headers())
        return self._handle_response(response)

    def get_environment_by_id(self, env_id: str) -> Dict[str, Any]:
        """Get specific environment status by ID"""
        environments = self.list_environments()
        if environments and "environments" in environments:
            for env in environments["environments"]:
                if env.get("id") == env_id or env.get("env_id") == env_id:
                    return {"active": True, "environment": env}
        return {"active": False, "environment": None}


    def stop_environment(self, env_id: str = None) -> Dict[str, Any]:
        """Stop user environment (graceful shutdown). If env_id is provided, stop that environment."""
        try:
            result = self.delete_environment(env_id=env_id)
            if result.get("status") == "deleted":
                return {"status": "success", "message": "Environment stopped successfully"}
            else:
                return {"status": "error", "message": f"Failed to stop environment: {result.get('message', 'Unknown error')}"}
        except Exception as e:
            return {"status": "error", "message": f"Error stopping environment: {str(e)}"}
    

    def restart_environment(self, env_id: str = None, config: Optional[Dict] = None) -> Dict[str, Any]:
        """Restart user environment. If env_id is provided, restart that environment."""
        try:
            # First, try to stop the environment
            delete_result = self.delete_environment(env_id=env_id)
            if delete_result.get("status") != "deleted":
                return {"status": "error", "message": f"Failed to stop environment for restart: {delete_result.get('message', 'Unknown error')}"}
            
            # Wait a bit for cleanup
            import time
            time.sleep(3)  # Longer pause for proper cleanup
            
            # Create new environment
            create_result = self.create_environment(config)
            if create_result.get("status") in ["created", "existing"]:
                return {"status": "success", "message": "Environment restarted successfully", "environment": create_result}
            else:
                return {"status": "error", "message": f"Failed to create new environment: {create_result.get('message', 'Unknown error')}"}
                
        except Exception as e:
            return {"status": "error", "message": f"Error restarting environment: {str(e)}"}
    
    # Storage Management Methods
    def list_user_storages(self) -> Dict[str, Any]:
        """List all storage buckets for the current user"""
        try:
            response = self.session.get(
                f"{self.base_url}/storage/list",
                headers=self._get_headers(),
                verify=False
            )
            return self._handle_response(response)
        except Exception as e:
            return {"status": "error", "message": f"Error listing storages: {str(e)}"}
    
    def get_storage_details(self, storage_id: str) -> Dict[str, Any]:
        """Get detailed information about a specific storage bucket"""
        try:
            response = self.session.get(
                f"{self.base_url}/storage/{storage_id}",
                headers=self._get_headers(),
                verify=False
            )
            return self._handle_response(response)
        except Exception as e:
            return {"status": "error", "message": f"Error getting storage details: {str(e)}"}
    
    def create_storage(self, storage_class: str = "STANDARD", custom_name: str = None) -> Dict[str, Any]:
        """Create a new storage bucket"""
        try:
            data = {"storage_class": storage_class}
            if custom_name:
                data["custom_name"] = custom_name
            
            response = self.session.post(
                f"{self.base_url}/storage/create",
                json=data,
                headers=self._get_headers(),
                verify=False
            )
            return self._handle_response(response)
        except Exception as e:
            return {"status": "error", "message": f"Error creating storage: {str(e)}"}
    
    def delete_storage(self, storage_id: str, force: bool = False) -> Dict[str, Any]:
        """Delete a storage bucket"""
        try:
            body = {"operation": "delete", "force": force}
            response = self.session.request(
                method="DELETE",
                url=f"{self.base_url}/storage/{storage_id}",
                json=body,
                headers=self._get_headers(),
                verify=False
            )
            return self._handle_response(response)
        except Exception as e:
            return {"status": "error", "message": f"Error deleting storage: {str(e)}"}

# Global API client instance
api_client = CMBClusterAPIClient()

File: nextjs-frontend/src/lib/api-client.ts
import axios, { AxiosInstance, AxiosResponse, AxiosError, InternalAxiosRequestConfig } from 'axios';
import { getSession, signOut } from 'next-auth/react';
import { getApiUrlAsync } from '@/lib/env-validator';
import type { 
  Environment, 
  StorageItem, 
  UserFile, 
  ApiResponse, 
  EnvironmentConfig,
  UserEnvVar,
  ApplicationImage,
  ApplicationImageRequest,
  UserWithRole
} from '@/types';

// Production configuration with enhanced security
const API_CONFIG = {
  // baseURL is now set dynamically at runtime in the request interceptor
  timeout: 30000, // 30 seconds
  retries: 3,
  retryDelay: 1000, // 1 second
};

class CMBClusterAPIClient {
  private api: AxiosInstance;
  private isRefreshing = false;
  private refreshPromise: Promise<string | null> | null = null;

  constructor() {
    this.api = axios.create({
      timeout: API_CONFIG.timeout,
      headers: {
        'Content-Type': 'application/json',
        'X-Requested-With': 'XMLHttpRequest', // CSRF protection
      },
      validateStatus: (status) => status < 500, // Don't throw on 4xx errors
      withCredentials: true, // Include cookies for CSRF protection
      // Add timeout for requests
      timeout: API_CONFIG.timeout,
    });

    // Request interceptor to add auth token and security headers
    this.api.interceptors.request.use(
      async (config: InternalAxiosRequestConfig) => {
        try {
          // Dynamically set the baseURL from runtime configuration to solve build-time env var issues
          config.baseURL = await getApiUrlAsync();

          // Get backend token (either from session or throw error)
          const backendToken = await this.getBackendToken();
          
          if (backendToken) {
            if (!config.headers) {
              config.headers = {} as any;
            }
            config.headers.Authorization = `Bearer ${backendToken}`;
          }
        } catch (error) {
          // If token retrieval fails, reject the request with helpful error
          return Promise.reject({
            isAuthError: true,
            message: error instanceof Error ? error.message : 'Authentication failed',
            config,
          });
        }
        
        // Add security headers
        config.headers = config.headers || {};
        config.headers['X-Client-Version'] = '1.0.0';
        config.headers['X-Client-Type'] = 'nextjs-frontend';
        
        return config;
      },
      (error) => {
        return Promise.reject(error);
      }
    );

    // Response interceptor for enhanced error handling and token refresh
    this.api.interceptors.response.use(
      (response: AxiosResponse) => {
        // Check for security warnings in response headers
        const securityWarning = response.headers['x-security-warning'];
        if (securityWarning) {
          console.warn('Security warning from API:', securityWarning);
        }
        
        return response;
      },
      async (error: AxiosError | any) => {
        // Handle authentication errors from request interceptor
        if (error.isAuthError) {
          console.error('Authentication error:', error.message);
          return Promise.reject(new Error(error.message));
        }

        const originalRequest = error.config;
        
        if (error.response?.status === 401) {
          // Authentication failed - sign out and redirect to login
          signOut({ callbackUrl: '/auth/signin' });
          
        } else if (error.response?.status === 403) {
          
        } else if (error.response?.status === 429) {
          
        } else if (error.response && error.response.status >= 500) {
          
        } else if (error.code === 'ECONNABORTED') {
          
        } else if (error.code === 'ERR_NETWORK') {
        }
        
        return Promise.reject(error);
      }
    );
  }

  /**
   * Get backend JWT token from NextAuth session
   */
  private async getBackendToken(): Promise<string | null> {
    try {
      const session = await getSession();
      
      // If no session, can't get token
      if (!session) {
        return null;
      }
      
   
      // If session has backend token, use it
      if (session.accessToken) {
   
        return session.accessToken;
      }
      
      // If session exists but no backend token, the initial token exchange probably failed

      throw new Error('Authentication incomplete - please try logging out and logging in again');
    } catch (error) {
      throw error;
    }
  }

  /**
   * Get headers for API requests
   */
  private async getHeaders(): Promise<Record<string, string>> {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'X-Client-Version': '1.0.0',
      'X-Client-Type': 'nextjs-frontend',
    };

    const backendToken = await this.getBackendToken();
    if (backendToken) {
      headers.Authorization = `Bearer ${backendToken}`;
    }

    return headers;
  }

  private async refreshToken(): Promise<string | null> {
    if (this.isRefreshing) {
      return this.refreshPromise;
    }

    this.isRefreshing = true;
    this.refreshPromise = this.performTokenRefresh();

    try {
      const result = await this.refreshPromise;
      return result;
    } finally {
      this.isRefreshing = false;
      this.refreshPromise = null;
    }
  }

  private async performTokenRefresh(): Promise<string | null> {
    try {
      // In NextAuth, token refresh happens automatically
      // We just need to get a fresh session
      const session = await getSession();
      return session?.accessToken || null;
    } catch (error) {
      console.error('Token refresh request failed:', error);
      return null;
    }
  }

  private async handleResponse<T>(response: AxiosResponse<T>): Promise<T> {
    return response.data;
  }

  private handleError(error: any): ApiResponse {
    console.error('API Error:', error);
    console.error('Error details:', {
      message: error.message,
      code: error.code,
      response: error.response?.data,
      status: error.response?.status,
      statusText: error.response?.statusText
    });
    
    const errorResponse: ApiResponse = {
      success: false,
      status: 'error',
      environments: [],
      storages: [],
      data: null,
      error: 'An error occurred'
    };

    if (error.response?.data) {
      errorResponse.error = error.response.data.detail || error.response.data.message || error.response.data.error || 'An error occurred';
      errorResponse.message = errorResponse.error;
    } else if (error.code === 'ECONNABORTED') {
      errorResponse.error = 'Request timeout - please try again';
      errorResponse.message = errorResponse.error;
    } else if (error.code === 'ERR_NETWORK') {
      errorResponse.error = 'Network error - please check your connection';
      errorResponse.message = errorResponse.error;
    } else if (error.message) {
      errorResponse.error = error.message;
      errorResponse.message = error.message;
    } else {
      errorResponse.error = 'An unexpected error occurred';
      errorResponse.message = errorResponse.error;
    }
    
    return errorResponse;
  }

  // Health check with enhanced error handling
  async healthCheck(): Promise<ApiResponse> {
    try {
      const response = await this.api.get('/health');
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  // Application management
  async listApplications(): Promise<ApiResponse<ApplicationImage[]>> {
    try {
      const response = await this.api.get('/applications');
      const data = await this.handleResponse(response);
      
      // Backend returns direct array, wrap in ApiResponse format
      if (Array.isArray(data)) {
        return {
          status: 'success',
          data: data,
          applications: data
        };
      }
      
      return data;
    } catch (error) {
      return this.handleError(error);
    }
  }

  async getApplication(applicationId: string): Promise<ApiResponse<ApplicationImage>> {
    try {
      const response = await this.api.get(`/applications/${applicationId}`);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  // Environment management
  async createEnvironment(config?: EnvironmentConfig): Promise<ApiResponse<Environment>> {
    try {
      const data = {
        cpu_limit: 2.0,
        memory_limit: "4Gi",
        storage_size: "20Gi",
        ...config
      };

      const response = await this.api.post('/environments', data);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async listEnvironments(): Promise<ApiResponse<Environment[]>> {
    try {
      const response = await this.api.get('/environments/list');
      const data = await this.handleResponse(response);
      
      // Backend returns { environments: [...] }
      if (data && typeof data === 'object' && 'environments' in data) {
        return {
          status: 'success',
          environments: data.environments,
          data: data.environments
        };
      }
      
      // Fallback for direct array response
      if (Array.isArray(data)) {
        return {
          status: 'success',
          environments: data,
          data: data
        };
      }
      
      return {
        status: 'error',
        message: 'Invalid response format',
        environments: []
      };
    } catch (error) {
      console.error('listEnvironments error:', error);
      return this.handleError(error);
    }
  }

  async getEnvironmentStatus(): Promise<ApiResponse<Environment[]>> {
    try {
      const response = await this.api.get('/environments');
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async getEnvironmentById(envId: string): Promise<ApiResponse<Environment>> {
    try {
      
      // Try direct API endpoint first
      try {
        const response = await this.api.get(`/environments/${envId}/info`);
        const data = await this.handleResponse(response);
        
        if (data && data.environment) {
          return {
            status: 'success',
            data: data.environment,
            environment: data.environment
          };
        }
      } catch (directError) {
      }

      // Fallback to list approach
      const listResponse = await this.listEnvironments();
      if (listResponse.environments) {
        const env = listResponse.environments.find(
          e => e.id === envId || e.env_id === envId
        );
        if (env) {
          return {
            status: 'success',
            data: env,
            environment: env
          };
        }
      }
      
      return {
        status: 'error',
        message: 'Environment not found'
      };
    } catch (error) {
      console.error('getEnvironmentById error:', error);
      return this.handleError(error);
    }
  }

  async restartEnvironment(envId?: string, config?: EnvironmentConfig): Promise<ApiResponse> {
    try {
      
      // Get current environment configuration before deleting
      let currentConfig = config;
      if (!currentConfig && envId) {
        try {
          const envResponse = await this.getEnvironmentById(envId);
          if (envResponse.environment?.resource_config) {
            currentConfig = {
              cpu_limit: envResponse.environment.resource_config.cpu_limit,
              memory_limit: envResponse.environment.resource_config.memory_limit,
              storage_size: envResponse.environment.resource_config.storage_size
            };
          }
        } catch (e) {
        }
      }

      // Stop environment first
      const deleteResponse = await this.deleteEnvironment(envId);
      
      if (deleteResponse.status !== 'deleted' && deleteResponse.status !== 'success') {
        throw new Error(`Failed to stop environment: ${deleteResponse.message}`);
      }

      // Wait a moment for cleanup
      await new Promise(resolve => setTimeout(resolve, 3000));

      // Create new environment with previous config
      const createResponse = await this.createEnvironment(currentConfig);
      
      if (createResponse.status === 'created' || createResponse.status === 'existing') {
        return {
          status: 'success',
          message: 'Environment restarted successfully',
          environment: createResponse.environment
        };
      }

      throw new Error(`Failed to create new environment: ${createResponse.message}`);
    } catch (error) {
      console.error('Restart environment error:', error);
      throw error;
    }
  }

  async stopEnvironment(envId?: string): Promise<ApiResponse> {
    try {
      const result = await this.deleteEnvironment(envId);
      if (result.status === 'deleted' || result.status === 'success') {
        return {
          status: 'success',
          message: 'Environment stopped successfully'
        };
      }
      throw new Error(result.message || 'Failed to stop environment');
    } catch (error) {
      console.error('Stop environment error:', error);
      throw error;
    }
  }

  async deleteEnvironment(envId?: string): Promise<ApiResponse> {
    try {
      const params = envId ? { env_id: envId } : {};
      const response = await this.api.delete('/environments', { params });
      return await this.handleResponse(response);
    } catch (error) {
      throw error;
    }
  }

  async sendHeartbeat(): Promise<ApiResponse> {
    try {
      const response = await this.api.post('/environments/heartbeat');
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  // Storage management
  async listUserStorages(): Promise<ApiResponse<StorageItem[]>> {
    try {
      const response = await this.api.get('/storage');
      const data = await this.handleResponse(response);
      
      // Backend returns { storages: [...], total_count: number, usage_stats: {...} }
      if (data && typeof data === 'object' && 'storages' in data) {
        return {
          status: 'success',
          storages: data.storages,
          data: data.storages,
          total_count: data.total_count,
          usage_stats: data.usage_stats
        };
      }
      
      // Fallback for direct array response
      if (Array.isArray(data)) {
        return {
          status: 'success',
          storages: data,
          data: data
        };
      }
      
      return {
        status: 'error',
        message: 'Invalid response format',
        storages: []
      };
    } catch (error) {
      console.error('listUserStorages error:', error);
      return this.handleError(error);
    }
  }

  async getStorageDetails(storageId: string): Promise<ApiResponse<StorageItem>> {
    try {
      const response = await this.api.get(`/storage/${storageId}`);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async createStorageBucket(storageClass: string = 'STANDARD', customName?: string): Promise<ApiResponse<StorageItem>> {
    try {
      const data: any = { storage_class: storageClass };
      if (customName) {
        data.custom_name = customName;
      }

      const response = await this.api.post('/storage', data);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async deleteStorage(storageId: string, force: boolean = false): Promise<ApiResponse> {
    try {
      const response = await this.api.delete(`/storage/${storageId}`, {
        data: { operation: 'delete', force }
      });
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  // User environment variables
  async getUserEnvVars(): Promise<ApiResponse<Record<string, string>>> {
    try {
      const response = await this.api.get('/user-env-vars');
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async setUserEnvVar(key: string, value: string): Promise<ApiResponse> {
    try {
      const response = await this.api.post('/user-env-vars', { key, value });
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async deleteUserEnvVar(key: string): Promise<ApiResponse> {
    try {
      const response = await this.api.delete(`/user-env-vars/${key}`);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  // User file management
  async listUserFiles(): Promise<ApiResponse<UserFile[]>> {
    try {
      const response = await this.api.get('/user-files');
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async uploadUserFile(
    file: File, 
    fileType: string, 
    envVarName: string, 
    containerPath?: string
  ): Promise<any> {
    try {
 
      // Validate file is JSON
      if (!file.name.toLowerCase().endsWith('.json')) {
        throw new Error('Only JSON files are allowed');
      }

      // Validate file size (1MB limit)
      if (file.size > 1024 * 1024) {
        throw new Error('File size must be less than 1MB');
      }

      // Validate required fields
      if (!fileType) {
        throw new Error('File type is required');
      }
      if (!envVarName?.trim()) {
        throw new Error('Environment variable name is required');
      }

      
      const formData = new FormData();
      formData.append('file', file, file.name);
      formData.append('file_type', fileType);
      formData.append('environment_variable_name', envVarName.trim());
      
      if (containerPath?.trim()) {
        formData.append('container_path', containerPath.trim());
      }

      for (let [key, value] of formData.entries()) {
        if (value instanceof File) {
        } else {
        }
      }

      
      const response = await this.api.post('/user-files/upload', formData, {
        headers: {
          'Content-Type': undefined, // Let axios set multipart/form-data with boundary
        },
      });
      
    
      const result = await this.handleResponse(response);
   
      return result;
    } catch (error) {

      if ((error as any)?.response) {
       
      }
      if ((error as any)?.request) {
       
      }
      throw error;
    }
  }

  async getUserFile(fileId: string): Promise<ApiResponse<UserFile>> {
    try {
      const response = await this.api.get(`/user-files/${fileId}`);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async updateUserFile(fileId: string, updates: Partial<UserFile>): Promise<ApiResponse<UserFile>> {
    try {
      const response = await this.api.put(`/user-files/${fileId}`, updates);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async deleteUserFile(fileId: string): Promise<ApiResponse> {
    try {
      const response = await this.api.delete(`/user-files/${fileId}`);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async downloadUserFile(fileId: string): Promise<ApiResponse> {
    try {
      const response = await this.api.get(`/user-files/${fileId}/download`);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  // Storage file operations
  async listStorageFiles(storageId: string, prefix?: string): Promise<any> {
    try {
      const params: any = {};
      if (prefix) params.prefix = prefix;
      const response = await this.api.get(`/storage/${storageId}/files`, { params });
      return response.data;
    } catch (error) {
      return this.handleError(error);
    }
  }

  async uploadFileToStorage(storageId: string, file: File, path?: string): Promise<any> {
    try {
     

      // Validate inputs
      if (!storageId) {
        throw new Error('Storage ID is required');
      }
      if (!file) {
        throw new Error('File is required');
      }

      
      const formData = new FormData();
      formData.append('file', file, file.name);
      
      const params: any = {};
      if (path) {
        params.path = path;
      }

      for (let [key, value] of formData.entries()) {
        if (value instanceof File) {
        } else {
        }
      }
     
      
      const response = await this.api.post(`/storage/${storageId}/upload`, formData, {
        params: params,
        headers: {
          'Content-Type': undefined, // Let axios set multipart/form-data with boundary
        },
      });
      
   
      const result = await this.handleResponse(response);
     
      return result;
    } catch (error) {
      console.error('=== STORAGE FILE UPLOAD ERROR ===');
      console.error('Storage upload error caught:', error);
      if ((error as any)?.response) {
        console.error('Response status:', (error as any).response.status);
        console.error('Response data:', (error as any).response.data);
        console.error('Response headers:', (error as any).response.headers);
      }
      if ((error as any)?.request) {
        console.error('Request details:', (error as any).request);
      }
      console.error('=== END STORAGE FILE UPLOAD ERROR ===');
      throw error;
    }
  }

  async downloadFileFromStorage(storageId: string, fileName: string): Promise<any> {
    try {
      
      // For path parameters, we need to encode each segment separately
      // to preserve slashes between folders but encode special characters
      const pathSegments = fileName.split('/').map(segment => encodeURIComponent(segment));
      const encodedPath = pathSegments.join('/');
      const url = `/storage/${storageId}/download/${encodedPath}`;
      
      const response = await this.api.get(url, {
        responseType: 'blob',
        timeout: 30000, // 30 second timeout for downloads
      });
      
     
      
      return response;
    } catch (error) {
      console.error('API download error:', error);
      throw error;
    }
  }

  async deleteFileFromStorage(storageId: string, fileName: string): Promise<ApiResponse> {
    try {
      const response = await this.api.delete(`/storage/${storageId}/files`, {
        data: { file_path: fileName }
      });
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  // Activity log
  async getActivityLog(limit: number = 50): Promise<ApiResponse> {
    try {
      const response = await this.api.get('/activity', { params: { limit } });
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  // ============ ADMIN API METHODS ============

  // Application Management (Admin)
  async createApplication(app: ApplicationImageRequest): Promise<ApiResponse<ApplicationImage>> {
    try {
      const response = await this.api.post('/admin/applications', app);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async updateApplication(id: string, app: Partial<ApplicationImageRequest>): Promise<ApiResponse<ApplicationImage>> {
    try {
      const response = await this.api.put(`/admin/applications/${id}`, app);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async deleteApplication(id: string): Promise<ApiResponse> {
    try {
      const response = await this.api.delete(`/admin/applications/${id}`);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async getAdminApplications(): Promise<ApiResponse<ApplicationImage[]>> {
    try {
      const response = await this.api.get('/admin/applications');
      const data = await this.handleResponse(response);
      
      // Admin endpoints return direct arrays, wrap in ApiResponse format
      if (Array.isArray(data)) {
        return {
          status: 'success',
          data: data,
          applications: data
        };
      }
      
      return data;
    } catch (error) {
      return this.handleError(error);
    }
  }

  // Application creation/update with image upload
  async createApplicationWithImage(formData: FormData): Promise<ApiResponse<ApplicationImage>> {
    try {
      
      // Use axios to post FormData with proper headers
      const response = await this.api.post('/admin/applications-with-image', formData, {
        headers: {
          'Content-Type': 'multipart/form-data'
        }
      });
      
      return await this.handleResponse(response);
    } catch (error) {
      console.error('Error in createApplicationWithImage:', error);
      return this.handleError(error);
    }
  }

  async updateApplicationWithImage(id: string, formData: FormData): Promise<ApiResponse<ApplicationImage>> {
    try {
      
      // Use axios to put FormData with proper headers  
      const response = await this.api.put(`/admin/applications-with-image/${id}`, formData, {
        headers: {
          'Content-Type': 'multipart/form-data'
        }
      });
      
      return await this.handleResponse(response);
    } catch (error) {
      console.error('Error in updateApplicationWithImage:', error);
      return this.handleError(error);
    }
  }

  // User Management (Admin)
  async getAllUsers(): Promise<ApiResponse<UserWithRole[]>> {
    try {
      const response = await this.api.get('/admin/users');
      const data = await this.handleResponse(response);
      
      // Admin endpoints return direct arrays, wrap in ApiResponse format
      if (Array.isArray(data)) {
        return {
          status: 'success',
          data: data,
          users: data
        };
      }
      
      return data;
    } catch (error) {
      return this.handleError(error);
    }
  }

  async updateUserRole(userId: string, newRole: 'user' | 'admin' | 'researcher', reason?: string): Promise<ApiResponse> {
    try {
      const response = await this.api.post(`/admin/users/${userId}/role`, {
        new_role: newRole,
        reason
      });
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async deactivateUser(userId: string): Promise<ApiResponse> {
    try {
      const response = await this.api.post(`/admin/users/${userId}/deactivate`);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async getUserRoleHistory(userId: string): Promise<ApiResponse> {
    try {
      const response = await this.api.get(`/admin/users/${userId}/role-history`);
      return await this.handleResponse(response);
    } catch (error) {
      return this.handleError(error);
    }
  }
}

// Global API client instance
export const apiClient = new CMBClusterAPIClient();
export default apiClient;
